pkg/coscheduling/coscheduling.go

```diff
@@ -95,294 +83,364 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
-	current := bound + 1
-
-	if current < minAvailable {
-		klog.V(3).Infof("The count of podGroup %v/%v/%v is not up to minAvailable(%d) in Permit: current(%d)",
-			pod.Namespace, podGroupName, pod.Name, minAvailable, current)
-		// TODO Change the timeout to a dynamic value depending on the size of the `PodGroup`
-		return framework.NewStatus(framework.Wait, ""), time.Duration(*cs.args.PermitWaitingTimeSeconds) * time.Second
+// RemovePod is called by the framework while trying to evaluate the impact
+// of removing podToRemove from the node while scheduling podToSchedule.
+func (cs *Coscheduling) RemovePod(ctx context.Context, _ *framework.CycleState,
+	podToSchedule *v1.Pod, podToRemove *v1.Pod, _ *framework.NodeInfo) *framework.Status {
+	if err := cs.pgMgr.PreemptRemovePod(ctx, podToSchedule, podToRemove); err != nil {
+		if klog.V(5).Enabled() {
+			klog.Error(err)
+		}
+		return framework.NewStatus(framework.Unschedulable, err.Error())
 	}
+	klog.V(5).Infof("CoschedulingExtension pass remove pod %v", podToRemove.Name)
+	return framework.NewStatus(framework.Success, "")
+}
 
-	klog.V(3).Infof("The count of PodGroup %v/%v/%v is up to minAvailable(%d) in Permit: current(%d)",
-		pod.Namespace, podGroupName, pod.Name, minAvailable, current)
-	cs.frameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
-		if waitingPod.GetPod().Namespace == namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
-			klog.V(3).Infof("Permit allows the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
-			waitingPod.Allow(cs.Name())
+// Permit is the functions invoked by the framework at "Permit" extension point.
+func (cs *Coscheduling) Permit(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	var pgs *schecache.PodGroupSchedulingStatus
+	waitTime := *cs.scheduleTimeout
+	ready, fullName, err := cs.pgMgr.Permit(ctx, pod, nodeName)
+	if err != nil {
+		if err == util.ErrorWaiting {
+			klog.Infof("Pod: %v/%v is waiting to be scheduled to node: %v", pod.Namespace, pod.Name, nodeName)
+			return framework.NewStatus(framework.Wait, ""), waitTime
 		}
-	})
+		// For pod not belongs to any groups
+		if err == util.ErrorNotMatched {
+			return framework.NewStatus(framework.Success, ""), 0
 
-	return framework.NewStatus(framework.Success, ""), 0
+		}
+		klog.Infof("cs.pgMgr.Permit error %v", err)
+		return framework.NewStatus(framework.Unschedulable, err.Error()), waitTime
+	}
+	if pgs = cs.pgMgr.PodGroupStatusCache().Get(fullName); pgs == nil {
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, "PodGroup not found"), waitTime
+	}
+	if wait := util.GetWaitTimeDuration(pgs.PodGroup, cs.scheduleTimeout); wait != 0 {
+		waitTime = wait
+	}
+	klog.V(5).Infof("Pod requires pgName %v", fullName)
+	if ready {
+		go cs.sendStartScheduleSignal(fullName)
```

Comment:
Do we need a goroutine here? cs.sendStartScheduleSignal() is not a blocking call.